sql/tests: TestRandomSyntaxSelect failed [inverted columns have no StableID]

sql/tests.TestRandomSyntaxSelect [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11467153?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11467153?buildTab=artifacts#/) on release-23.1.9-rc @ [ba4b4e7aef964ee55699ba1437424c6436cc76a4](https://github.com/cockroachdb/cockroach/commits/ba4b4e7aef964ee55699ba1437424c6436cc76a4):

Random syntax error:

```
    rsg_test.go:882: Crash detected: server panic: pq: internal error: inverted columns have no StableID
```
Query:

```
        SELECT DEFAULT IS NULL , * FROM [ 42 ( 229 ) AS PASSWORD ( TREAT , POSITION , ENUM , OVERLAY , MULTIPOLYGONM , EXISTS , NOCREATELOGIN ) ] @ CHARACTERISTICS WITH ORDINALITY WHERE NOT TIMETZ 'string';
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #109398 sql/tests: TestRandomSyntaxSelect failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSelect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30901

sql/tests.TestRandomSyntaxSelect [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11516862?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11516862?buildTab=artifacts#/) on release-23.1.9-rc @ [49148609ccafb4dd3bc547ef3aaa10c95b4f8dbf](https://github.com/cockroachdb/cockroach/commits/49148609ccafb4dd3bc547ef3aaa10c95b4f8dbf):


```
    rsg_test.go:855: 2m0s of 5m0s: 16210 executions, 336 successful
    rsg_test.go:855: 2m5s of 5m0s: 16811 executions, 351 successful
    rsg_test.go:855: 2m10s of 5m0s: 17391 executions, 364 successful
    rsg_test.go:855: 2m15s of 5m0s: 18023 executions, 380 successful
    rsg_test.go:855: 2m20s of 5m0s: 18612 executions, 393 successful
    rsg_test.go:855: 2m25s of 5m0s: 19162 executions, 405 successful
    rsg_test.go:855: 2m30s of 5m0s: 19692 executions, 424 successful
    rsg_test.go:855: 2m35s of 5m0s: 20278 executions, 437 successful
    rsg_test.go:855: 2m40s of 5m0s: 20845 executions, 455 successful
    rsg_test.go:855: 2m45s of 5m0s: 21419 executions, 464 successful
    rsg_test.go:855: 2m50s of 5m0s: 21980 executions, 476 successful
    rsg_test.go:855: 2m55s of 5m0s: 22499 executions, 491 successful
    rsg_test.go:855: 3m0s of 5m0s: 23016 executions, 501 successful
    rsg_test.go:855: 3m5s of 5m0s: 23593 executions, 517 successful
    rsg_test.go:855: 3m10s of 5m0s: 24115 executions, 529 successful
    rsg_test.go:855: 3m15s of 5m0s: 24637 executions, 541 successful
    rsg_test.go:855: 3m20s of 5m0s: 25172 executions, 550 successful
    rsg_test.go:855: 3m25s of 5m0s: 25695 executions, 564 successful
    rsg_test.go:855: 3m30s of 5m0s: 26262 executions, 573 successful
    rsg_test.go:855: 3m35s of 5m0s: 26806 executions, 588 successful
    rsg_test.go:855: 3m40s of 5m0s: 27342 executions, 601 successful
    rsg_test.go:855: 3m45s of 5m0s: 27876 executions, 614 successful
    rsg_test.go:855: 3m50s of 5m0s: 28414 executions, 630 successful
    rsg_test.go:855: 3m55s of 5m0s: 28925 executions, 644 successful
    rsg_test.go:855: 4m0s of 5m0s: 29454 executions, 657 successful
    rsg_test.go:855: 4m5s of 5m0s: 29952 executions, 674 successful
    rsg_test.go:855: 4m10s of 5m0s: 30469 executions, 689 successful
    rsg_test.go:855: 4m15s of 5m0s: 30987 executions, 701 successful
    rsg_test.go:855: 4m20s of 5m0s: 31499 executions, 715 successful
    rsg_test.go:855: 4m25s of 5m0s: 32013 executions, 728 successful
    rsg_test.go:855: 4m30s of 5m0s: 32540 executions, 740 successful
    rsg_test.go:855: 4m35s of 5m0s: 33059 executions, 751 successful
    rsg_test.go:855: 4m40s of 5m0s: 33620 executions, 758 successful
    rsg_test.go:855: 4m45s of 5m0s: 34100 executions, 767 successful
    rsg_test.go:855: 4m50s of 5m0s: 34599 executions, 773 successful
    rsg_test.go:855: 4m55s of 5m0s: 35116 executions, 791 successful
    rsg_test.go:855: 5m0s of 5m0s: 35598 executions, 798 successful
    rsg_test.go:211: Context cancelled while executing: "SELECT PLACEHOLDER ASC FROM [ SHOW CREATE EXTERNAL CONNECTION error ] AS ident ( ident , ident ) WHERE ARRAY ( ( ( TABLE error ) ) )"
    rsg_test.go:211: Context cancelled while executing: "SELECT * , - UNIQUE ( error AS AUTHORIZATION FROM ident "
    rsg_test.go:211: Context cancelled while executing: "SELECT UNIQUE ( error COLLATE ORDINALITY TYPECAST POINT error ARRAY IS OF ( POLYGON error ARRAY , INTERVAL ARRAY [ 92 ] [ error ) AS CROSS FROM [ -390 ( ) ident ( ident ) ] WITH ORDINALITY COMMIT , ONLY INDEX_AFTER_ORDER_BY_BEFORE_AT @ [ -360 ] RELOCATE WHERE ROW ( ) OVERLAPS ROW ( ) COLLATE ABORT IS NOT NAN = UNIQUE ( error IS NOT DISTINCT FROM DEFAULT IS FALSE COLLATE ident"
    rsg_test.go:211: Context cancelled while executing: "SELECT ROW ( ) OVERLAPS ( ( ) OVERLAPS ROW ( ) , ) FROM ident "
    rsg_test.go:211: Context cancelled while executing: "SELECT UNIQUE ( error ISNULL NOT_EQUALS UNIQUE ( error TYPECAST POLYGON error ARRAY [ 435 ] [ error ident FROM LATERAL [ FUNCTION -54 ] ( ) BOX2D NATURAL JOIN ( WITH RECURSIVE ident ( IFERROR , SUBSTRING ) AS MATERIALIZED ( RESUME SCHEDULES error ) TABLE error ) WITH ORDINALITY ident ( INOUT ) WHERE CBRT ( ) OVERLAPS ( ) ISNULL"
    rsg_test.go:211: Context cancelled while executing: "SELECT DEFAULT NOT IN ( ( ( error FOR NO KEY UPDATE FOR SHARE OF FAMILY , INDEX_BEFORE_PAREN . MAXVALUE , INDEX_BEFORE_NAME_THEN_PAREN . USER , INDEX_BEFORE_PAREN . CHARACTERISTICS , MATERIALIZED . ident . FAMILY NOWAIT FOR KEY SHARE OF NUMERIC SKIP LOCKED FOR UPDATE OF FAMILY . LEAST , ident . INNER . NOLOGIN , COMPLETIONS . ARRAY . REAL , ident NOWAIT ) ) IS NOT UNKNOWN FROM [ 282 ( ) ident ( USE , ident , ALWAYS ) ] @ SELECT AS EXPERIMENTAL_RELOCATE ( ident ) WHERE - UNIQUE ( error IS NOT NULL IS NOT FALSE"
    rsg_test.go:211: Context cancelled while executing: "SELECT ( ) OVERLAPS ROW ( ) GEOMETRY FROM LATERAL STDIN ( * ) AS ident ( ident , VIEWCLUSTERSETTING , VARBIT ) WHERE INTERVAL 'string' [ ROW ( ) OVERLAPS ( ) : ]"
    rsg_test.go:211: Context cancelled while executing: "SELECT OPERATOR ( ident . * ) DEFAULT IS NOT NULL AS DESC , * FROM ident . VARBIT @ CAST , [ ( error ORDER BY INDEX_AFTER_ORDER_BY_BEFORE_AT NOTHING_AFTER_RETURNING . ident . POINT @ TREAT DESC ] WITH ORDINALITY AS ident , LATERAL [ FUNCTION 488 ] ( * ) ident ( INCREMENTAL_LOCATION , CAPABILITY ) WHERE UNIQUE ( error IS NOT UNKNOWN IS OF ( INTERVAL YEAR ARRAY [ 102 ] , ident ARRAY [ 469 ] [ error , ident ARRAY , DEALLOCATE . FAMILY ARRAY , SCANS ARRAY ) IS NOT FALSE"
    rsg_test.go:891: 35606 executions, 798 successful
    rsg_test.go:899: cannot parse output of Format: sql="SELECT EXISTS ( WITH OUT ( ident ) AS MATERIALIZED ( ALTER DEFAULT PRIVILEGES FOR ALL ROLES IN SCHEMA ident . ident REVOKE GRANT OPTION FOR GRANT ON FUNCTIONS FROM ident ) , IFERROR AS NOT MATERIALIZED ( DROP EXTERNAL CONNECTION error ) ( ( WITH ident AS MATERIALIZED ( DROP SCHEDULE error ) TABLE [ 227 ( ) CANCELQUERY ( PRECEDING ) ] @ [ 27 ] AS WITHOUT ( BOOLEAN ) ORDER BY INDEX_AFTER_ORDER_BY_BEFORE_AT FAMILY . NEW_DB_NAME . FAMILY @ ident DESC ) ) ) AS LOCAL FROM ident ", formattedSQL="SELECT EXISTS (WITH \"out\" (ident) AS MATERIALIZED (ALTER DEFAULT PRIVILEGES FOR ALL ROLES IN SCHEMA ident.ident REVOKE GRANT OPTION FOR Kind(0) ON FUNCTIONS  FROM ident), \"iferror\" AS NOT MATERIALIZED (DROP EXTERNAL CONNECTION 'error') ((WITH ident AS MATERIALIZED (DROP SCHEDULES VALUES (error)) TABLE [227() AS cancelquery (preceding)]@[27] AS without (\"boolean\") ORDER BY INDEX \"family\".new_db_name.\"family\"@ident DESC))) AS local FROM ident": at or near "(": syntax error
    rsg_test.go:339: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSelect2487993115
--- FAIL: TestRandomSyntaxSelect (301.15s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #109549 sql/tests: TestRandomSyntaxSelect failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSelect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/optbuilder/testdata/table_ref
+++ b/pkg/sql/opt/optbuilder/testdata/table_ref
@@ -1,0 +1,20 @@
+exec-ddl
+CREATE TABLE t (
+  k INT PRIMARY KEY,
+  j JSON,
+  INVERTED INDEX (j)
+)
+----
+
+build
+SELECT * FROM [53 AS foo]
+----
+project
+ ├── columns: k:1!null j:2
+ └── scan t [as=foo]
+      └── columns: k:1!null j:2 crdb_internal_mvcc_timestamp:3 tableoid:4
+
+build
+SELECT * FROM [53(5) AS foo]
+----
+error (42703): column [5] does not exist

--- a/pkg/sql/opt/optbuilder/util.go
+++ b/pkg/sql/opt/optbuilder/util.go
@@ -694,7 +694,9 @@ func resolveNumericColumnRefs(tab cat.Table, columns []tree.ColumnID) (ordinals 
 		cnt := tab.ColumnCount()
 		for ord < cnt {
 			col := tab.Column(ord)
-			if col.ColID() == cat.StableID(c) && col.Visibility() != cat.Inaccessible {
+			// NOTE: Inverted columns cannot be referenced.
+			if col.Kind() != cat.Inverted && col.ColID() == cat.StableID(c) &&
+				col.Visibility() != cat.Inaccessible {
 				break
 			}
 			ord++

